### PR TITLE
Unbox earlier, avoid second pass

### PR DESCRIPTION
**Pull request by alainfrisch**
This PR change the way unboxing of floats and boxed integers is done.  See http://caml.inria.fr/mantis/view.php?id=7022 for details.  This improves compile-time complexity and enables more unboxing.  It will also hopefully interact better with other optimizations to come, as the one in http://caml.inria.fr/mantis/view.php?id=7017 .

Mandatory micro-benchmark:

``` ocaml
let () =
  let module M = Nativeint in
  for k = 1 to 10000 do
    let x = ref (M.of_int 1) in
    for i = 1 to 100000 do
      let y =
        if i mod 2 = 0 then !x
        else M.add !x (M.of_int 1)
      in
      x := M.add !x y;
    done;
  done
```

Trunk: 3.4s
Branch: 1.2s

### CODE REVIEW

**Comment by chambart** [comment]
You could have propagated an environment. Cmmgen is already sufficiently hard to follow, please don't make it worse.

**Comment by alainfrisch** [comment]
An environment, with all the plumbing required to pass it everywhere, would certainly make the patch harder to read.  I'm undecided about the final version, though.

**Comment by damiendoligez** [comment]
@alainfrisch: complexity of the patch is a secondary consideration, what's important is readability of the resulting code.

**Comment by alainfrisch** [comment]
Ok, I've pushed an extra commit which passes the info through an explicit environment.  I've used a generic name `env` so that, if later changes require to pass different kinds of info, we won't need to rename the argument or add new ones.

**Comment by mshinwell** [comment]
Pierre and I are in support of this patch.
We had a similar transformation in the "un-ANF" stage of flambda.  This was needed so that intermediate "let"s emitted by earlier flambda passes didn't inhibit unboxing in Cmmgen.  However, if Cmmgen can be made to cope with this (and we think this patch will do the job), that seems like a better solution.

**Comment by alainfrisch** [comment]
Committed to trunk, rev 16527.

**Comment by mshinwell** [comment]
(I may not have been clear: Pierre and I have not yet read this patch in detail for correctness.)

**Comment by bluddy** [comment]
Alain, what about int un-tagging? Is it possible to apply the same logic to untag ints so that mathematical operations are cheaper?

**Comment by alainfrisch** [comment]
A similar framework (in particular the extra environment passed down during translation) could be used to keep int identifier untagged, but I suspect the gain would be quite small.

**Comment by bluddy** [comment]
@alainfrisch , I don't think the gain would be small. Try to compile a simple calculation loop (such as incrementing an integer) in ocaml vs c, or just about any other compiled language (even haskell). You'll see a massive performance difference, because ocaml keeps the integer tagged. It's great for the GC and generic functions, but really bad for any arithmetic operations. If we want to be competitive (which I think we do), I believe we need to have an untagged int mode where tight loops are converted to untagged integers locally.

**Comment by alainfrisch** [comment]
Incrementing a tagged integer amounts to adding 2.  I doubt this is significantly slower than adding 1.  Are you sure the slowdown comes from the tagging?

**Comment by bluddy** [comment]
OK good point. It was probably an operation other than incrementing. Maybe I was multiplying by a number in a loop. Whatever it was, I compared the assembly outputs and performance, and the difference in performance was astounding.

**Comment by mshinwell** [comment]
bluddy: Can you please provide us with your test cases / assembly outputs?

**Comment by alainfrisch** [comment]
The discussion diverged a little bit.  @bluddy If you have example of code that would benefit from eliminating more tagging/untagging, please open a ticket on Mantis.
